### PR TITLE
feat:[CDS-70330]: Resource for encrypting and decrypting TF plan for harness SM

### PR DESCRIPTION
**Pull request by sourabh-awashti**
## Describe your changes
Changes are corresponding to support for apis in manager to encrypt or decrypt TF plan through harness SM.
Tests Done:-
1. TF and Terragrunt pipeline execution with encryption and decryption on manager through apis
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48706)

### CODE REVIEW

**Review by mohitg0795 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48706)* status: 8 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/resources/EncryptionDecryptionResource.java` line 44 at r1](https://reviewable.io/reviews/harness/harness-core/48706#-NWvlH5XBbQjeei1rVBP:-NWvlH5XBbQjeei1rVBQ:b-qhnp5r) ([raw file](https://github.com/harness/harness-core/blob/35e144c1ed851c33eb7be8cea960008f93ad4840/400-rest/src/main/java/software/wings/resources/EncryptionDecryptionResource.java#L44)):*
> ```Java
>   @DelegateAuth
>   @POST
>   @Path("/encrypt-HarnessSM-secret")
> ```

path is case insensitive

**Review by mohitg0795 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48706)* status: 8 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/service/impl/EncryptionDecryptionServiceImpl.java` line 31 at r1](https://reviewable.io/reviews/harness/harness-core/48706#-NWvlhzU64-RQKoNBQBZ:-NWvlhzU64-RQKoNBQB_:bgt5x6l) ([raw file](https://github.com/harness/harness-core/blob/35e144c1ed851c33eb7be8cea960008f93ad4840/400-rest/src/main/java/software/wings/service/impl/EncryptionDecryptionServiceImpl.java#L31)):*
> ```Java
> @OwnedBy(CDP)
> public class EncryptionDecryptionServiceImpl implements EncryptionDecryptionService {
>   @Inject protected KmsEncryptorsRegistry kmsEncryptorsRegistry;
> ```

protected for some reason?

**Comment by mohitg0795** [comment]
*[`400-rest/src/main/java/software/wings/service/impl/EncryptionDecryptionServiceImpl.java` line 41 at r1](https://reviewable.io/reviews/harness/harness-core/48706#-NWvloy-0YJvPb2Uj4xa:-NWvloy-0YJvPb2Uj4xb:bvdnp4l) ([raw file](https://github.com/harness/harness-core/blob/35e144c1ed851c33eb7be8cea960008f93ad4840/400-rest/src/main/java/software/wings/service/impl/EncryptionDecryptionServiceImpl.java#L41)):*
> ```Java
>   }
> 
>   private EncryptedRecord encryptKmsSecret(String value, EncryptionConfig encryptionConfig) {
> ```

nit: move private methods at the last of the file for better readability

File: 400-rest/src/main/java/software/wings/resources/EncryptionDecryptionResource.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.MediaType;
+import lombok.extern.slf4j.Slf4j;
+import retrofit.http.Body;
+
+@Api("/encrypt-decrypt")
+@Path("/encrypt-decrypt")
+@Produces("application/json")
```
Comment (yogesh-chauhan) [inline 400-rest/src/main/java/software/wings/resources/EncryptionDecryptionResource.java:36]: can we have a better name like DelegateManagerClient etc

File: 930-delegate-tasks/src/main/java/io/harness/managerclient/EncryptionDecryptionHarnessSMClient.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import io.harness.security.encryption.EncryptedRecordData;
+
+import javax.ws.rs.Consumes;
+import okhttp3.RequestBody;
+import retrofit2.Call;
+import retrofit2.http.Body;
+import retrofit2.http.POST;
+import retrofit2.http.Query;
+
+@TargetModule(HarnessModule._420_DELEGATE_AGENT)
```
Comment (yogesh-chauhan) [inline 930-delegate-tasks/src/main/java/io/harness/managerclient/EncryptionDecryptionHarnessSMClient.java:24]: target is not needed

File: 930-delegate-tasks/src/main/java/io/harness/managerclient/EncryptionDecryptionHarnessSMClient.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+public interface EncryptionDecryptionHarnessSMClient {
+  @POST("encrypt-decrypt/encrypt-HarnessSM-secret")
+  @Consumes({"application/x-protobuf"})
+  Call<RestResponse<EncryptedRecordData>> encryptHarnessSMSecret(
+      @Query("accountId") String accountId, @Body RequestBody content);
```
Comment (yogesh-chauhan) [inline 930-delegate-tasks/src/main/java/io/harness/managerclient/EncryptionDecryptionHarnessSMClient.java:29]: we should mark accountId as not null

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]

File: 400-rest/src/main/java/software/wings/service/impl/EncryptionDecryptionServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+@OwnedBy(CDP)
+public class EncryptionDecryptionServiceImpl implements EncryptionDecryptionService {
+  @Inject protected KmsEncryptorsRegistry kmsEncryptorsRegistry;
+  @Inject private SecretManagerConfigService secretManagerConfigService;
+
+  @Override
+  public EncryptedRecordData encryptData(byte[] content, String accountId) {
+    String value = new String(CHARSET.decode(ByteBuffer.wrap(encodeBase64ToByteArray(content))).array());
+    SecretManagerConfig secretManagerConfig = secretManagerConfigService.getDefaultSecretManager(accountId);
```
Comment (yogesh-chauhan) [inline 400-rest/src/main/java/software/wings/service/impl/EncryptionDecryptionServiceImpl.java:37]: this should be getGlobalSecretManager(..)

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]

File: 400-rest/src/main/java/software/wings/service/impl/EncryptionDecryptionServiceImpl.java (hunk 4 above)
Comment (sourabh-awashti) [reply 400-rest/src/main/java/software/wings/service/impl/EncryptionDecryptionServiceImpl.java:37]: done

File: 930-delegate-tasks/src/main/java/io/harness/managerclient/EncryptionDecryptionHarnessSMClient.java (hunk 2 above)
Comment (sourabh-awashti) [reply 930-delegate-tasks/src/main/java/io/harness/managerclient/EncryptionDecryptionHarnessSMClient.java:24]: done

File: 930-delegate-tasks/src/main/java/io/harness/managerclient/EncryptionDecryptionHarnessSMClient.java (hunk 3 above)
Comment (sourabh-awashti) [reply 930-delegate-tasks/src/main/java/io/harness/managerclient/EncryptionDecryptionHarnessSMClient.java:29]: done

**Comment by sourabh-awashti** [comment]
*[`400-rest/src/main/java/software/wings/resources/EncryptionDecryptionResource.java` line 36 at r1](https://reviewable.io/reviews/harness/harness-core/48706#gh-1214276923:-NX9Q2HJCSYJ_E61jYkn:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/35e144c1ed851c33eb7be8cea960008f93ad4840/400-rest/src/main/java/software/wings/resources/EncryptionDecryptionResource.java#L36)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

can we have a better name like DelegateManagerClient etc
</blockquote></details>

Done.

**Comment by sourabh-awashti** [comment]
*[`400-rest/src/main/java/software/wings/service/impl/EncryptionDecryptionServiceImpl.java` line 31 at r1](https://reviewable.io/reviews/harness/harness-core/48706#-NWvlhzU64-RQKoNBQBZ:-NX9Q4iO2LdTjU9JjTiq:bx9f9oe) ([raw file](https://github.com/harness/harness-core/blob/35e144c1ed851c33eb7be8cea960008f93ad4840/400-rest/src/main/java/software/wings/service/impl/EncryptionDecryptionServiceImpl.java#L31)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

protected for some reason?
</blockquote></details>

will remove it

**Comment by mohitg0795** [comment]
*[`400-rest/src/main/java/software/wings/resources/DelegateManagerEncryptionDecryptionHarnessSMResource.java` line 51 at r2](https://reviewable.io/reviews/harness/harness-core/48706#-NXGj0dXEqQQboDM7tkw:-NXGj0dXEqQQboDM7tkx:blqzax2) ([raw file](https://github.com/harness/harness-core/blob/f9a01802b8c0f1ac3bc3474301ee20a87f2493cd/400-rest/src/main/java/software/wings/resources/DelegateManagerEncryptionDecryptionHarnessSMResource.java#L51)):*
> ```Java
>     } catch (Exception e) {
>       log.error(format("Unable to encrypt the content for harness secret manager for account %s", accountId), e);
>       return new RestResponse<>(EncryptedRecordData.builder().build());
> ```

this should be error response

**Comment by mohitg0795** [comment]
*[`400-rest/src/main/java/software/wings/resources/DelegateManagerEncryptionDecryptionHarnessSMResource.java` line 65 at r2](https://reviewable.io/reviews/harness/harness-core/48706#-NXGj8xg7k3GT4iMtenQ:-NXGj8xg7k3GT4iMtenR:b-2guby3) ([raw file](https://github.com/harness/harness-core/blob/f9a01802b8c0f1ac3bc3474301ee20a87f2493cd/400-rest/src/main/java/software/wings/resources/DelegateManagerEncryptionDecryptionHarnessSMResource.java#L65)):*
> ```Java
>     } catch (Exception e) {
>       log.error(format("Unable to decrypt the content for harness secret manager for account %s", accountId), e);
>       return new RestResponse<>(DecryptedRecord.builder().decryptedValue(new char[0]).build());
> ```

same here

**Comment by mohitg0795** [comment]
*[`400-rest/src/main/java/software/wings/service/impl/DelegateManagerEncryptionDecryptionHarnessSMServiceImpl.java` line 37 at r2](https://reviewable.io/reviews/harness/harness-core/48706#-NXGjWHL7hHCcunPYNc-:-NXGjWHL7hHCcunPYNc0:b-ooavak) ([raw file](https://github.com/harness/harness-core/blob/f9a01802b8c0f1ac3bc3474301ee20a87f2493cd/400-rest/src/main/java/software/wings/service/impl/DelegateManagerEncryptionDecryptionHarnessSMServiceImpl.java#L37)):*
> ```Java
>   @Override
>   public EncryptedRecordData encryptData(String accountId, byte[] content) {
>     String value = new String(CHARSET.decode(ByteBuffer.wrap(encodeBase64ToByteArray(content))).array());
> ```

this should be done inside a util

**Review by mohitg0795 — DISMISSED** [review]
> Reviewed 13 of 13 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48706)* status: 13 file(s) to review. 1 change requests. 4. unresolved discussions. 1 of 2 approvals obtained.

**Review by yogesh-chauhan — DISMISSED** [review]
> Reviewed 13 of 13 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48706)* status: :shipit: complete! ✓

**Comment by sourabh-awashti** [comment]
trigger uts

**Comment by sourabh-awashti** [comment]
*[`400-rest/src/main/java/software/wings/resources/DelegateManagerEncryptionDecryptionHarnessSMResource.java` line 51 at r2](https://reviewable.io/reviews/harness/harness-core/48706#-NXGj0dXEqQQboDM7tkw:-NXL_WEF8SyNLaSWncVd:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/f9a01802b8c0f1ac3bc3474301ee20a87f2493cd/400-rest/src/main/java/software/wings/resources/DelegateManagerEncryptionDecryptionHarnessSMResource.java#L51)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

this should be error response
</blockquote></details>

Done.

**Comment by sourabh-awashti** [comment]
*[`400-rest/src/main/java/software/wings/resources/DelegateManagerEncryptionDecryptionHarnessSMResource.java` line 65 at r2](https://reviewable.io/reviews/harness/harness-core/48706#-NXGj8xg7k3GT4iMtenQ:-NXL_XYT4e2WHEe6MCo7:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/f9a01802b8c0f1ac3bc3474301ee20a87f2493cd/400-rest/src/main/java/software/wings/resources/DelegateManagerEncryptionDecryptionHarnessSMResource.java#L65)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

same here
</blockquote></details>

Done.

**Review by mohitg0795 — APPROVED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48706)* status: 1 file(s) to review. 1 of 2 approvals obtained.

**Comment by sourabh-awashti** [comment]
trigger uts

**Comment by sourabh-awashti** [comment]
trigger messagecheck

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48706)* status: :shipit: complete! ✓

**Comment by yogesh-chauhan** [comment]
trigger review

**Comment by sourabh-awashti** [comment]
trigger review

**Pull request merged**